README.md
Using beanstalkd made easy
==========================

azuki makes it easy to create tools using
[beanstalkd](http://kr.github.io/beanstalkd/) tubes for asynchronous execution.
From queueing parallelizing quick shell hacks to integrating asynchronous
execution with your django application, azuki lets you do it all.

CLI beanstalk inspection
------------------------

The main utility, `azuki`, makes it easy to inspect and modify beanstalk queues
from the command line.

```
$ azuki tubes [-v]
$ azuki stats
$ azuki stats default
$ azuki stats 123
```

The stats command will give you basic statistics, such as the total amount of
connections, jobs and commands.  If you specify a tube, such as `default`, you
will see the statistics of that tube, and if you specify a job ID, you will see
its details and content.

The tubes command shows a list of tubes. Add the -v parameter to also get a
summary of all processes and jobs for all queues.

Peeking at tubes
----------------

```
$ azuki peek-delayed default
$ azuki peek-ready default
$ azuki peek-buried default
$ azuki peek 123
```

With the first three commands above you can look at the first delayed, ready or
buried job in a tube. If you use the `--ask` option as well, you will be asked
whether you want to delete, bury or kick the job. The author is fond of `azuki
peek-buried --ask` to do cleanup of failed jobs.

The last command will peek at a specific job, given its id number.

Manipulating jobs and tubes
---------------------------

`azuki kick 100 default`

This would kick 100 jobs in the default queue, moving them from buried back to
ready.

```
azuki bury 123
azuki kick 123
```

If you want to bury or kick a specific job, use the commands above.

Queueing shell commands
-----------------------
Azuki was originally written to easily parallelize shell jobs. Of course [gnu
parallel](http://www.gnu.org/software/parallel/) can be used if you just want
parallelization, but I also needed to be able to adjust parallelism, pause
everything and add jobs while the application was running. In short, I needed
to use a queue, and beanstalk is my queue of choice.

The original task was to update the iLO firmware of thousands of HP servers in
an automated, yet controlled way. To do this, first one puts jobs in the tube:

```
for ilo in $(<ilos.txt); do
    echo $ilo | azuki put --ttr 600 ilo-firmware
done
```

This reads a list of items and schedules them one by one in the ilo-firmware
tube with a ttr of 600. You can also specify a delay with the `--delay`
argument.

To consume the items in the tube, you use `azuki foreach`.

`azuki foreach ilo-firmware -- xargs do_firmware_update`

The `foreach` command will execute a command you want to execute and feed the
job body on stdin. `do_firmware_update` wants this on the command line instead,
hence the use of `xargs`. To now parallelize this, we can simply run as many
instances of `azuki foreach` as we want in parallel, possibly using screen.

The `foreach` command looks at the exitcode of the commands it runs. If the
exitcode is 0, the job is considered succesfull and gets deleted. If the
exitcode is nonzero, the job gets buried.

At the end of my working day I wanted to pause the upgrading, but using
`ctrl-C` in the middle of a firmware upgrade is not a very good idea. Azuki to
the rescue!

`azuki pause 8640000 ilo-firmware`

This pauses the tube for 100 days. Already running firmware updates would
complete, but no new ones would be scheduled. 10 minutes later all is paused.
To unpause the tube, simply pause it again, but for 0 seconds.

Python API
----------
beanstalkd tubes already are fairly easy to use in python, using the beanstalkc
library. But fairly easy isn't easy enough, so azuki can make it easier for
you. You can simply decorate your functions to make them asynchronous.

example.py:
```
from azuki import beanstalk

@beanstalk('example-tube')
def hello(who):
    print "Hello, %s" % who
```

main.py:
```
import example

example.hello("world")
```

If you now run `python main.py`, you will notice that it does not output
anything. Instead it has serialized the arguments to `hello` and scheduled them
as a job in the `example-tube` tube.

To process the queued jobs, you run `azuki daemon example-tube`. This will take
items from the tube, import the `example` module and call the `hello` function
for real.

Rescheduling
------------
With beanstalk you can bury a job to tell beanstalk not to attempt the job
again until manually told to do so, but if you just want to delay execution of
a job, you can raise an `azuki.Reschedule` exception.

```
from azuki import beanstalk, Reschedule

@beanstalk('example-tube')
def process(task):
    if not am_ready_for(task):
        raise Reschedule(120)
    do_task(task)
```

Django API
----------
The downside of scheduling things in beanstalk queues, is that argument to
function calls must be serialized. Azuki uses json serialization, so anything
that is not json-serializable, cannot be used as an argument.

Except django model instances, as azuki recognizes them and handles them
specially. That means that for example queueing mails instead of sending them
directly works:

models.py:
```
from django.contrib.auth import User
from django.core.mail import send_mail
from azuki import beanstalk

class Message(models.Model):
    recipient = models.ForeignKey(User)
    subject = models.CharField("Subject", max_length=128)
    text = models.TextField("Message text")

    @beanstalk('send-mail')
    def send(self):
        send_mail(self.subject, self.text, 'webmaster@localhost', [self.recipient])
```

If you now create a message and call `send()`, the message is not sent, but
only added to the `send-mail` tube. You can again use `azuki daemon` to process
this tube and actually send the mails, possibly even on a different machine
altogether.

Author
------
(c) 2014, Dennis Kaarsemaker <dennis@kaarsemaker.net>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


debian/changelog
azuki (1.2-1) wily; urgency=medium

  * New upstream release 

 -- Dennis Kaarsemaker <dennis@kaarsemaker.net>  Sun, 04 Oct 2015 23:22:21 +0200

azuki (1.1-1) wily; urgency=medium

  * New upstream release 

 -- Dennis Kaarsemaker <dennis@kaarsemaker.net>  Sun, 04 Oct 2015 23:00:44 +0200

azuki (1.0-1) vivid; urgency=medium

  * New upstream release

 -- Dennis Kaarsemaker <dennis@kaarsemaker.net>  Sun, 12 Apr 2015 09:16:42 +0200

azuki (0.9.2-1) utopic; urgency=medium

  * New upstream release

 -- Dennis Kaarsemaker <dennis@seahawk>  Thu, 18 Sep 2014 14:13:33 +0200

azuki (0.9.1-1) trusty; urgency=medium

  * Initial version

 -- Dennis Kaarsemaker <dennis@kaarsemaker.net>  Tue, 16 Sep 2014 11:05:16 +0200


test/setup.sh
. ./sharness.sh
export PYTHONPATH="$SHARNESS_BUILD_DIRECTORY/lib:$SHARNESS_BUILD_DIRECTORY/test/python:$PYTHONPATH"
export PATH="$SHARNESS_BUILD_DIRECTORY/bin:$PATH"

export test_tube="azuki-test-$this_test"

if [ x"$PYTHON" = x ]; then
    PYTHON=python
fi

$PYTHON -c 'import django' 2>/dev/null && test_set_prereq DJANGO

alias azuki="$PYTHON $SHARNESS_BUILD_DIRECTORY/bin/azuki"

clean_tubes() {
    for tube in $(azuki tubes); do
        case $tube in
            azuki-test-*)
                submit_test_job
                while azuki stats $tube | grep -q 'Delayed: *[1-9]'; do
                    echo delete | azuki peek-delayed --ask $test_tube > /dev/null
                done
                azuki kick 10000 $tube
                azuki foreach $tube /bin/true > /dev/null
                ;;
        esac
    done
}

job_counter=0
submit_test_job() {
    job_counter=$(expr $job_counter + 1)
    echo "test job $job_counter" | azuki put "$@" "$test_tube"
}

test_python() {
    ( cd $SHARNESS_TEST_DIRECTORY/python && $PYTHON -munittest $1 )
}


test/python/t_plain.py
from azuki import beanstalk
import os
import sys
import unittest

@beanstalk(os.environ['test_tube'])
def hello(arg):
    print("Hello, %s" % arg)
    sys.exit(0)

@beanstalk(os.environ['test_tube'], priority=1)
def fast_hello(arg):
    print("Hello, important %s" % arg)

class AzukiTest(unittest.TestCase):
    def test_azuki(self):
        self.assertEqual(type(hello("world")), int)
        self.assertEqual(type(fast_hello("world")), int)

    def test_serializer(self):
        self.assertRaises(TypeError, lambda: hello(self))


test/python/t_django.py
import os
import unittest

os.environ['DJANGO_SETTINGS_MODULE'] = 'azk.settings'
from azk.azu.models import Azu

class DjangoTest(unittest.TestCase):
    def test_django(self):
        obj = Azu.objects.create(name="django")
        self.assertEqual(type(obj.greet()), int)


lib/azuki/daemon.py
try:
    import beanstalkc
except (ImportError, SyntaxError):
    # Hey, we're on python 3, let's load our fixed version!
    import azuki.beanstalkc as beanstalkc
import azuki
import imp
import logging
import sys
import json
import traceback

try:
    from django.core.exceptions import ImproperlyConfigured
    from django.conf import settings
    from django.db.models import get_model
    from django.db import connections
    import django
    if hasattr(django, 'setup'):
        django.setup()
except ImportError:
    django = False
except ImproperlyConfigured:
    django = False

class Daemon(object):
    def __init__(self, beanstalk):
        self.beanstalk = beanstalk
        self.logger = logging.getLogger("azuki")
        self.main_modules = {}

    def run(self):
        azuki.running_azuki_daemon = True
        self.connect()
        while True:
            job = self.fetch()
            try:
                self.handle(job)
                job.delete()
            except azuki.Reschedule:
                e = sys.exc_info()[1]
                stats = job.stats()
                self.bs.use(stats['tube'])
                self.bs.put(job.body, delay=e.delay, priority=stats['pri'], ttr=stats['ttr'])
                job.delete()
            except Exception:
                self.job_error(job, tb=traceback.format_exc())
            finally:
                if django:
                    # Close all DB connections so we don't hang/crash later
                    for connection in connections:
                        connections[connection].close()

    def watch(self, tube):
        azuki.all_tubes[self.beanstalk].add(tube)
    
    def job_error(self, job, message='', tb=None):
        self.logger.error("Error processing job %d from tube %s: %s" % (job.jid, job.stats()['tube'], message))
        if tb:
            for line in tb.split('\n'):
                self.logger.error(line)
        job.bury()

    def handle(self, job):
        try:
            job.data = json.loads(job.body)
        except ValueError:
            return self.job_error(job, "Could not decode job")
        handler = job.data.get('handler', None)
        if not hasattr(self, 'handle_' + handler):
            return self.job_error(job, "Unknown handler '%s'" % handler)
        getattr(self, 'handle_' + handler)(job)

    # TODO: handle ImportError, AttributeError
    def handle_function(self, job):
        if job.data['module'] == '__main__':
            if job.data['file'] not in self.main_modules:
                with open(job.data['file']) as fd:
                    dwb = sys.dont_write_bytecode
                    sys.dont_write_bytecode = True
                    self.main_modules[job.data['file']] = imp.load_module('azuki.fake_main', fd, job.data['file'], ('', 'r', imp.PY_SOURCE))
                    sys.dont_write_bytecode = dwb
            module = self.main_modules[job.data['file']]
        else:
            if job.data['module'] not in sys.modules:
                __import__(job.data['module'])
            module = sys.modules[job.data['module']]
        getattr(module, job.data['function'])(*job.data['args'], **job.data['kwargs'])

    # TODO: hanlde model not found, instance not found, AttributeError
    def handle_django(self, job):
        if not django:
            return self.job_error(job, "Unable to handle django jobs")
        model = get_model(job.data['app'], job.data['model'])
        instance = model.objects.get(pk=job.data['pk'])
        self.logger.info("Calling %s.%s.%s, pk %d" % (job.data['app'], job.data['model'], job.data['method'], job.data['pk']))
        getattr(instance, job.data['method'])(*job.data['args'], **job.data['kwargs'])

    def connect(self):
        self.bs = beanstalkc.Connection(**azuki.beanstalks[self.beanstalk])
        for tube in azuki.all_tubes[self.beanstalk]:
            self.bs.watch(tube)
    
    def fetch(self):
        self.logger.info("Waiting for job")
        try:
            return self.bs.reserve()
        except Exception:
            self.logger.error("Connection to beanstalk failed: %s, reconnecting" % str(sys.exc_info()[1]))
            self.connect()
            return self.bs.reserve()


lib/azuki/beanstalkc.py
#!/usr/bin/env python
"""beanstalkc - A beanstalkd Client Library for Python"""

__license__ = '''
Copyright (C) 2008-2014 Andreas Bolka

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
'''

__version__ = '0.4.0'

import logging
import socket
import sys


DEFAULT_HOST = 'localhost'
DEFAULT_PORT = 11300
DEFAULT_PRIORITY = 2 ** 31
DEFAULT_TTR = 120


PY3 = sys.version_info[0] > 2
if PY3:
    b = lambda x: bytes(x, sys.getdefaultencoding())
    s = lambda x: x.decode(sys.getdefaultencoding())
else:
    b = lambda x: x
    s = lambda x: x


class BeanstalkcException(Exception): pass
class UnexpectedResponse(BeanstalkcException): pass
class CommandFailed(BeanstalkcException): pass
class DeadlineSoon(BeanstalkcException): pass

class SocketError(BeanstalkcException):
    @staticmethod
    def wrap(wrapped_function, *args, **kwargs):
        try:
            return wrapped_function(*args, **kwargs)
        except socket.error:
            err = sys.exc_info()[1]
            raise SocketError(err)


class Connection(object):
    def __init__(self, host=DEFAULT_HOST, port=DEFAULT_PORT, parse_yaml=True,
                 connect_timeout=socket.getdefaulttimeout()):
        if parse_yaml is True:
            try:
                parse_yaml = __import__('yaml').load
            except ImportError:
                logging.error('Failed to load PyYAML, will not parse YAML')
                parse_yaml = False
        self._connect_timeout = connect_timeout
        self._parse_yaml = parse_yaml or (lambda x: x)
        self.host = host
        self.port = port
        self.connect()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def connect(self):
        """Connect to beanstalkd server."""
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._socket.settimeout(self._connect_timeout)
        SocketError.wrap(self._socket.connect, (self.host, self.port))
        self._socket.settimeout(None)
        self._socket_file = self._socket.makefile('rb')

    def close(self):
        """Close connection to server."""
        try:
            self._socket.sendall(b('quit\r\n'))
        except socket.error:
            pass
        try:
            self._socket.close()
        except socket.error:
            pass

    def reconnect(self):
        """Re-connect to server."""
        self.close()
        self.connect()

    def _interact(self, command, expected_ok, expected_err=[]):
        SocketError.wrap(self._socket.sendall, b(command))
        status, results = self._read_response()
        if status in expected_ok:
            return results
        elif status in expected_err:
            raise CommandFailed(command.split()[0], status, results)
        else:
            raise UnexpectedResponse(command.split()[0], status, results)

    def _read_response(self):
        line = SocketError.wrap(self._socket_file.readline)
        if not line:
            raise SocketError()
        response = s(line).split()
        return response[0], response[1:]

    def _read_body(self, size):
        body = SocketError.wrap(self._socket_file.read, size)
        SocketError.wrap(self._socket_file.read, 2)  # trailing crlf
        if size > 0 and not body:
            raise SocketError()
        return s(body)

    def _interact_value(self, command, expected_ok, expected_err=[]):
        return self._interact(command, expected_ok, expected_err)[0]

    def _interact_job(self, command, expected_ok, expected_err, reserved=True):
        jid, size = self._interact(command, expected_ok, expected_err)
        body = self._read_body(int(size))
        return Job(self, int(jid), body, reserved)

    def _interact_yaml(self, command, expected_ok, expected_err=[]):
        size, = self._interact(command, expected_ok, expected_err)
        body = self._read_body(int(size))
        return self._parse_yaml(body)

    def _interact_peek(self, command):
        try:
            return self._interact_job(command, ['FOUND'], ['NOT_FOUND'], False)
        except CommandFailed:
            return None

    # -- public interface --

    def put(self, body, priority=DEFAULT_PRIORITY, delay=0, ttr=DEFAULT_TTR):
        """Put a job into the current tube. Returns job id."""
        assert isinstance(body, str), 'Job body must be a str instance'
        jid = self._interact_value('put %d %d %d %d\r\n%s\r\n' % (
                                       priority, delay, ttr, len(body), body),
                                   ['INSERTED'],
                                   ['JOB_TOO_BIG', 'BURIED', 'DRAINING'])
        return int(jid)

    def reserve(self, timeout=None):
        """Reserve a job from one of the watched tubes, with optional timeout
        in seconds. Returns a Job object, or None if the request times out."""
        if timeout is not None:
            command = 'reserve-with-timeout %d\r\n' % timeout
        else:
            command = 'reserve\r\n'
        try:
            return self._interact_job(command,
                                      ['RESERVED'],
                                      ['DEADLINE_SOON', 'TIMED_OUT'])
        except CommandFailed:
            exc = sys.exc_info()[1]
            _, status, results = exc.args
            if status == 'TIMED_OUT':
                return None
            elif status == 'DEADLINE_SOON':
                raise DeadlineSoon(results)

    def kick(self, bound=1):
        """Kick at most bound jobs into the ready queue."""
        return int(self._interact_value('kick %d\r\n' % bound, ['KICKED']))

    def kick_job(self, jid):
        """Kick a specific job into the ready queue."""
        self._interact('kick-job %d\r\n' % jid, ['KICKED'], ['NOT_FOUND'])

    def peek(self, jid):
        """Peek at a job. Returns a Job, or None."""
        return self._interact_peek('peek %d\r\n' % jid)

    def peek_ready(self):
        """Peek at next ready job. Returns a Job, or None."""
        return self._interact_peek('peek-ready\r\n')

    def peek_delayed(self):
        """Peek at next delayed job. Returns a Job, or None."""
        return self._interact_peek('peek-delayed\r\n')

    def peek_buried(self):
        """Peek at next buried job. Returns a Job, or None."""
        return self._interact_peek('peek-buried\r\n')

    def tubes(self):
        """Return a list of all existing tubes."""
        return self._interact_yaml('list-tubes\r\n', ['OK'])

    def using(self):
        """Return the tube currently being used."""
        return self._interact_value('list-tube-used\r\n', ['USING'])

    def use(self, name):
        """Use a given tube."""
        return self._interact_value('use %s\r\n' % name, ['USING'])

    def watching(self):
        """Return a list of all tubes being watched."""
        return self._interact_yaml('list-tubes-watched\r\n', ['OK'])

    def watch(self, name):
        """Watch a given tube."""
        return int(self._interact_value('watch %s\r\n' % name, ['WATCHING']))

    def ignore(self, name):
        """Stop watching a given tube."""
        try:
            return int(self._interact_value('ignore %s\r\n' % name,
                                            ['WATCHING'],
                                            ['NOT_IGNORED']))
        except CommandFailed:
            return 1

    def stats(self):
        """Return a dict of beanstalkd statistics."""
        return self._interact_yaml('stats\r\n', ['OK'])

    def stats_tube(self, name):
        """Return a dict of stats about a given tube."""
        return self._interact_yaml('stats-tube %s\r\n' % name,
                                   ['OK'],
                                   ['NOT_FOUND'])

    def pause_tube(self, name, delay):
        """Pause a tube for a given delay time, in seconds."""
        self._interact('pause-tube %s %d\r\n' % (name, delay),
                       ['PAUSED'],
                       ['NOT_FOUND'])

    # -- job interactors --

    def delete(self, jid):
        """Delete a job, by job id."""
        self._interact('delete %d\r\n' % jid, ['DELETED'], ['NOT_FOUND'])

    def release(self, jid, priority=DEFAULT_PRIORITY, delay=0):
        """Release a reserved job back into the ready queue."""
        self._interact('release %d %d %d\r\n' % (jid, priority, delay),
                       ['RELEASED', 'BURIED'],
                       ['NOT_FOUND'])

    def bury(self, jid, priority=DEFAULT_PRIORITY):
        """Bury a job, by job id."""
        self._interact('bury %d %d\r\n' % (jid, priority),
                       ['BURIED'],
                       ['NOT_FOUND'])

    def touch(self, jid):
        """Touch a job, by job id, requesting more time to work on a reserved
        job before it expires."""
        self._interact('touch %d\r\n' % jid, ['TOUCHED'], ['NOT_FOUND'])

    def stats_job(self, jid):
        """Return a dict of stats about a job, by job id."""
        return self._interact_yaml('stats-job %d\r\n' % jid,
                                   ['OK'],
                                   ['NOT_FOUND'])


class Job(object):
    def __init__(self, conn, jid, body, reserved=True):
        self.conn = conn
        self.jid = jid
        self.body = body
        self.reserved = reserved

    def _priority(self):
        stats = self.stats()
        if isinstance(stats, dict):
            return stats['pri']
        return DEFAULT_PRIORITY

    # -- public interface --

    def delete(self):
        """Delete this job."""
        self.conn.delete(self.jid)
        self.reserved = False

    def release(self, priority=None, delay=0):
        """Release this job back into the ready queue."""
        if self.reserved:
            self.conn.release(self.jid, priority or self._priority(), delay)
            self.reserved = False

    def bury(self, priority=None):
        """Bury this job."""
        if self.reserved:
            self.conn.bury(self.jid, priority or self._priority())
            self.reserved = False

    def kick(self):
        """Kick this job alive."""
        self.conn.kick_job(self.jid)

    def touch(self):
        """Touch this reserved job, requesting more time to work on it before
        it expires."""
        if self.reserved:
            self.conn.touch(self.jid)

    def stats(self):
        """Return a dict of stats about this job."""
        return self.conn.stats_job(self.jid)


if __name__ == '__main__':
    import nose
    nose.main(argv=['nosetests', '-c', '.nose.cfg'])


lib/azuki/__init__.py
import logging
try:
    import beanstalkc
except (ImportError, SyntaxError):
    # Hey, we're on python 3, let's load our fixed version!
    import azuki.beanstalkc as beanstalkc
from functools import wraps
import json
from collections import defaultdict
import os

beanstalks = {'default': {'host': '127.0.0.1', 'port': 11300}}
running_azuki_daemon = False
all_tubes = defaultdict(set)

try:
    from django.core.exceptions import ImproperlyConfigured
    from django.conf import settings
    beanstalks = settings.BEANSTALK_SERVERS
except ImportError:
    django = False
except ImproperlyConfigured:
    django = False

def add_beanstalk(name, host, port=11300):
    beanstalks[name] = {'host': host, 'port': port}

def connect_cached(beanstalk, tube, __cache={}):
    bsc = beanstalks[beanstalk]
    if beanstalk not in __cache:
        __cache[beanstalk] = beanstalkc.Connection(**bsc)
    try:
        __cache[beanstalk].use(tube)
    except Exception:
        # Retry once
        __cache[beanstalk] = beanstalkc.Connection(**bsc)
        __cache[beanstalk].use(tube)
    return __cache[beanstalk]

def beanstalk(tube_or_func='default', beanstalk='default', priority=2147483648, delay=0, ttr=120):
    tube = 'default' if callable(tube_or_func) else tube_or_func
    def decorator(func, tube=tube, beanstalk=beanstalk, priority=priority, delay=delay, ttr=ttr):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if running_azuki_daemon:
                return func(*args, **kwargs)

            # For normal functions, serialize all arguments
            if not args or not hasattr(args[0], '_meta'):
                data = {
                    'handler':  'function',
                    'module':   func.__module__,
                    'function': func.__name__,
                    'args':     args,
                    'kwargs':   kwargs,
                }
                if func.__module__ == '__main__':
                    data['file'] = os.path.abspath(func.__code__.co_filename)
            else:
                self = args[0]; args=args[1:]
                data = {
                    'handler': 'django',
                    'app':     self._meta.app_label,
                    'model':   self._meta.object_name,
                    'method':  func.__name__,
                    'pk':      self.pk,
                    'args':    args,
                    'kwargs':  kwargs,
                }

            try:
                data = json.dumps(data)
            except TypeError:
                raise TypeError("Can only queue json-serializable arguments" + repr(data))

            bs = connect_cached(beanstalk, tube)
            return bs.put(data, priority=priority, delay=delay, ttr=ttr)
        wrapper.stats = lambda: connect_cached(beanstalk, tube).stats_tube(tube)
        all_tubes[beanstalk].add(tube)
        return wrapper

    if callable(tube_or_func):
        return decorator(tube_or_func, tube, beanstalk, priority, delay, ttr)
    return decorator

class Reschedule(Exception):
    def __init__(self, delay):
        self.delay = delay
